src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -102,95 +79,111 @@ BlkAllocStatus AppendBlkAllocator::alloc(blk_count_t nblks, const blk_alloc_hint
+    auto cpg = hs()->cp_mgr().cp_guard();
+    auto new_offset = blkid.blk_num() + blkid.blk_count();
+    auto cur_offset = m_commit_offset.load();
+    bool modified{true};
+    do {
+        if (cur_offset >= new_offset) {
+            // Already allocated
+            modified = false;
+            break;
+        }
+    } while (!m_commit_offset.compare_exchange_weak(cur_offset, new_offset));
+
+    if (modified) { m_is_dirty.store(true); }
+    return BlkAllocStatus::SUCCESS;
+}
 
-bool AppendBlkAllocator::is_blk_alloced_on_disk(BlkId const&, bool) const { return false; }
+// Given that AppendBlkAllocator uses highest of all offsets during the commit during recovery, we don't need to
+// individually recover each block.
+BlkAllocStatus AppendBlkAllocator::reserve_on_cache(BlkId const& blkid) {
+    auto new_offset = blkid.blk_num() + blkid.blk_count();
+    auto cur_offset = m_last_append_offset.load();
+    do {
+        if (cur_offset >= new_offset) { break; }
+    } while (!m_last_append_offset.compare_exchange_weak(cur_offset, new_offset));
+    return BlkAllocStatus::SUCCESS;
+}
 
-//
-// cp_flush doesn't need CPGuard as it is triggered by CPMgr which already handles the reference check;
-//
-// cp_flush should not block alloc/free;
-//
 void AppendBlkAllocator::cp_flush(CP* cp) {
-    const auto idx = cp->id() % MAX_CP_COUNT;
     // check if current cp's context has dirty buffer already
-    if (m_dirty_sb[idx].is_dirty) {
-        m_sb->last_append_offset = m_dirty_sb[idx].last_append_offset;
-        m_sb->freeable_nblks = m_dirty_sb[idx].freeable_nblks;
+    if (m_is_dirty.exchange(false)) {
+        m_sb->commit_offset = m_commit_offset.load();
+        m_sb->freeable_nblks = m_freeable_nblks.load();
 
         // write to metablk;
         m_sb.write();
-
-        // clear this dirty buff's dirty flag;
-        clear_dirty_offset(idx);
     }
 }
 
-// updating current cp's dirty buffer context;
-void AppendBlkAllocator::set_dirty_offset(const uint8_t idx) {
-    m_dirty_sb[idx].is_dirty = true;
-
-    m_dirty_sb[idx].last_append_offset = m_last_append_offset;
-    m_dirty_sb[idx].freeable_nblks = m_freeable_nblks;
-}
-
-// clearing current cp context's dirty flag;
-void AppendBlkAllocator::clear_dirty_offset(const uint8_t idx) { m_dirty_sb[idx].is_dirty = false; }
-
 //
 // free operation does:
 // 1. book keeping "total freeable" space
 // 2. if the blk being freed happens to be last block, move last_append_offset backwards accordingly;
 //
 void AppendBlkAllocator::free(const BlkId& bid) {
-    std::unique_lock lk(m_mtx);
-    auto cur_cp = hs()->cp_mgr().cp_guard();
-    const auto n = bid.blk_count();
-    m_freeable_nblks += n;
-    if (bid.blk_num() + n == m_last_append_offset) {
-        // we are freeing the the last blk id, let's rewind.
-        m_last_append_offset -= n;
+    // If we are freeing the last block, just move the offset back
+    // If we are freeing the last block, just move the offset back
```

Comment:
duplicate comment, I guess you mean if we are freeing in the middle, update m_freeable_nblks